lib/CPathfinder.cpp

```diff
@@ -1217,48 +1218,37 @@ int CPathfinderHelper::getMovementCost(
 		dt = hero->cb->getTile(dst);
 	}
 
-	/// TODO: by the original game rules hero shouldn't be affected by terrain penalty while flying.
-	/// Also flying movement only has penalty when player moving over blocked tiles.
-	/// So if you only have base flying with 40% penalty you can still ignore terrain penalty while having zero flying penalty.
 	int ret = hero->getTileCost(*dt, *ct, ti);
-	/// Unfortunately this can't be implemented yet as server don't know when player flying and when he's not.
-	/// Difference in cost calculation on client and server is much worse than incorrect cost.
-	/// So this one is waiting till server going to use pathfinder rules for path validation.
-
-	if(dt->blocked && ti->hasBonusOfType(Bonus::FLYING_MOVEMENT))
+    if(hero->boat != nullptr && dt->terType->isWater())
```

Comment:
I could be wrong, but this code may work incorrectly in some edge cases.
If hero is on land, but builds a path across favorable winds on water, this condition will evaluate to false.

Might even happen within single turn/move order - if hero has Admiral's Hat and can embark boat & reach favorable winds in a single turn.

If I am correct, can you either fix or add comment about such scenario?

Reply:
Looks like boat is really extra here. Another option will be to check if layer is SAIL. This will guarantee that the boat must exist.

Reply:
I believe EPathfindingLayer should be available somewhere upper in call stack

Reply:
I'll create a test map with AH and see what happens.

In one case the layer is available, in another one it seems not:

> it's also called from CGameHandler::moveHero() where those objects seem inaccessible